### PR TITLE
Minor improvements from user feedback

### PR DESCRIPTION
**Pull request by ryancarlson**
For follow up modal views
- Followup views will now adjust their height and y position to ensure that the keyboard doesn't hide the data input field that's being interacted with
- The next button will hop from one input field to the next.  It will change to 'Done' when on the last field
- The nav bar will not be shown when the follow up view modal dismisses

For the entire renderer:
- a single tap anywhere on the screen (besides an element of any kind) will hide the navbar if it's shown
- a double tap anywhere on the screen (besides an element of any kind will show the navbar.
- this fixes an annoyance that exists in production where the navbar will hide and immediately reshow itself on a single tap
- twenty taps exhibits some random behavior.  further investigation required..

### CODE REVIEW

**Comment by ryancarlson** [comment]
@michaelharro here's a three feedback related items from Ann i've fixed.  this one PR includes three small things.

File: GTViewController/Classes/Custom Classes/GTFollowupModalView.m [hunk 1]
```diff
@@ -70,16 +70,12 @@ - (instancetype)initFromElement:(TBXMLElement *)element withStyle:(GTPageStyle*)
...
-        CGFloat availableBufferSpace = totalVerticalSpace - totalUsedSpace;
-        topLeadingSpace =       availableBufferSpace * 0.15;
-        betweenElementsSpace =  (availableBufferSpace * 0.2) / (numLabels + numButtonPairs + numTextInputs);
-    }
+    CGFloat availableBufferSpace = totalVerticalSpace - totalUsedSpace;
+    topLeadingSpace =       availableBufferSpace * 0.15;
+    betweenElementsSpace =  (availableBufferSpace * 0.2) / (numLabels + numButtonPairs + numTextInputs);
     
     int currentY = topLeadingSpace;
+    int inputFieldTag = 0;
```
Comment (michaelharro) [inline GTViewController/Classes/Custom Classes/GTFollowupModalView.m]: Sorry to do this to you but you are going to get tag conflicts with inputFieldTag starting at zero. I give all elements on the page tags. The way we are avoiding tag conflicts is to give each type of element a decimal digit. I think buttons get 0-9, panels get 10-90, etc. You will need to look through the GTPage class and find a range that is free and claim it.

Comment (michaelharro) [inline GTViewController/Classes/Custom Classes/GTFollowupModalView.m]: I'm not sure if tag conflicts affect anything in the sdk but they are bad in our code because our animation logic depends on them not conflicting.

Comment (ryancarlson) [inline GTViewController/Classes/Custom Classes/GTFollowupModalView.m]: Okay.  I can fix that up.  Thanks for pointing it out.

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m [hunk 2]
```diff
@@ -100,4 +146,146 @@ - (void)transitionToThankYou {
...
+    
+    CGRect keyboardFrame = [self.view convertRect:keyboardEndFrame toView:nil];
+    CGRect inputFieldViewFrame = self.activeField.superview.frame;
+    
+    CGRect unobscuredByKeyboardFrame = self.view.frame;
+    unobscuredByKeyboardFrame.size.height -= keyboardFrame.size.height;
+    
+    // offset the height of the inputFieldView, b/c it is a label stacked on an input
+    // if just the frame is considered, then the label could show while the input is hidden
+    unobscuredByKeyboardFrame.size.height -= inputFieldViewFrame.size.height;
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: Its generally better practice to use the CGRectGetHeight(), CGRectGetMaxY(), etc methods

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m [hunk 3]
```diff
@@ -100,4 +146,146 @@ - (void)transitionToThankYou {
...
+
+- (void) animateViewWithVerticalDelta:(CGFloat)verticalDelta animationDuration:(NSTimeInterval) animationDuration animationCurve:(UIViewAnimationCurve)animationCurve {
+    // Get animation info from userInfo
+    CGRect newViewFrame = self.view.frame;
+    
+    [UIView beginAnimations:nil context:nil];
+    [UIView setAnimationDuration:animationDuration];
+    [UIView setAnimationCurve:animationCurve];
+    
+    newViewFrame.origin.y -= verticalDelta;
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: There are also methods that will do insets for you http://nshipster.com/cggeometry/

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m [hunk 4]
```diff
@@ -100,4 +146,146 @@ - (void)transitionToThankYou {
...
+    newViewFrame.size.height += verticalDelta;
+    self.view.frame = newViewFrame;
+    
+    [UIView commitAnimations];
+}
+
+
+- (void) resetViewPosition:(NSNotification *) notification {
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: Could you remove the extra spaces from this method`- (void)resetViewPosition:(NSNotification *)notification`

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m [hunk 5]
```diff
@@ -100,4 +146,146 @@ - (void)transitionToThankYou {
...
+        
+        CGFloat verticalDelta = [self calculateVerticalDeltaFromCurrentField:self.activeField.superview
+                                                                 toNextField:nextActiveField.superview];
+        
+        self.activeField = nextActiveField;
+        
+        [self animateViewWithVerticalDelta:verticalDelta animationDuration:0.3 animationCurve:UIViewAnimationCurveEaseOut];
+        
+        if (![nextActiveField.superview.superview viewWithTag:++nextTag]) {
+            [nextActiveField setReturnKeyType:UIReturnKeyDone];
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: What if there is only one field? Should there also be some sort of check like this when the keyboard is shown to make sure the return keys says done if there is only one field?

Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: jk. I just read you next method. i'm not used to a different ordering of these methods

File: GTViewController/Classes/Nav Classes/GTViewController.m [hunk 6]
```diff
@@ -1015,6 +1022,8 @@ - (void)hideNavToolbar {
         
     }
     
+    self.navToolbarIsShown		= NO;
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: why did you move this from the end of the animation to the beginning?

Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: It seemed like there was a race condition going on with button taps when I was debugging.  I could see justification for this going back, as I think I solved the problem another way.

File: GTViewController/Classes/Nav Classes/GTViewController.m [hunk 7]
```diff
@@ -345,7 +347,12 @@ - (void)pageEventDetected:(NSNotification *)notification {
         
         // dismiss a follow up modal if it is present and displayed
         if (self.followupViewController) {
-            [self dismissViewControllerAnimated:YES completion:nil];
+            self.bypassPresentingNavbar = YES;
+            __weak typeof(self) weakSelf = self;
+            
+            [self dismissViewControllerAnimated:YES completion:^{
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: I would call this method on self.followupViewController instead of self. dismissViewControllerAnimated either will dismiss the view controller the current view controller presented or if there is no presented view controller it will ask the parent view controller to dismiss the current view controller. calling the method on self.followupViewController means there is no chance of you accidentally dismissing the current view controller if the state got messed up. https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIViewController_Class/#//apple_ref/occ/instm/UIViewController/dismissViewControllerAnimated:completion:

File: GTViewController/Classes/Nav Classes/GTViewController.m [hunk 8]
```diff
@@ -1448,9 +1463,7 @@ - (void)processTap:(UITouch *)touch withEvent:(UIEvent *)event {
     
     if (touch.tapCount == 20) {
         [self fiftyTap];
-    }
-    
-    if (touch.tapCount == 1) {
+    } else if (touch.tapCount > 1) {
```
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: The PR comment also mentioned a double tap action. Where is that handled?

Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: Right here.. The tap count must be `> 1` instead of `== 1`

Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: sorry, i wasn't reading carefully enough

File: GTViewController/Classes/Custom Classes/GTFollowupModalView.m (hunk 1 above)
Comment (ryancarlson) [inline GTViewController/Classes/Custom Classes/GTFollowupModalView.m]: 👍

File: GTViewController/Classes/Nav Classes/GTViewController.m (hunk 6 above)
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: ok, only put it back if the race condition has been solved. I'm not sure it would matter if the nav bar got prematurely dismissed with this variable being set too early

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 2 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: 👍

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 3 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: 👍

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 4 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: I guess... 👍

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 5 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: 👍 I reordered the methods in a manner I think you would expect..

File: GTViewController/Classes/Nav Classes/GTViewController.m (hunk 7 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: Thanks for the pointer! 👍

File: GTViewController/Classes/Nav Classes/GTViewController.m (hunk 6 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTViewController.m]: I'll put it back and see what happens. 👍

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 3 above)
Comment (ryancarlson) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: So using these actually broke that functionality in a way I don't understand.  It stopped the animation from happening, no matter if I continued using `newViewFrame

**Comment by ryancarlson** [comment]
Thanks for the feedback!

File: GTViewController/Classes/Nav Classes/GTFollowupViewController.m (hunk 3 above)
Comment (michaelharro) [inline GTViewController/Classes/Nav Classes/GTFollowupViewController.m]: what was the code you used instead of this? Those methods create new CGRect structures, they don't act on self.view.frame itself. You will need to assign the newly created frame back to self.view.frame
That is the most common issue that comes up with this kind of animation.
